File: Benchmarks/Core/Systems/CameraSystem.cs
using Benchmarks.Core.Components;
using Benchmarks.Core.Entity;
using Benchmarks.Core.Scene;
using Benchmarks.Core.Systems.Contract;
using Benchmarks.Core.Types;

namespace Benchmarks.Core.Systems
{
    /// <summary>
    /// Sistema encargado de la camara.
    /// Replica el calculo de matrices de vista del engine.
    /// </summary>
    public class CameraSystem : ISystem
    {
        public Matrix ViewMatrix { get; private set; } = Matrix.Identity;
        public Matrix ProjectionMatrix { get; private set; } = Matrix.Identity;

        public override void InitializeSystem(EntityManager @entities)
        {
            EntityManager = @entities;
        }

        public override void UpdateSystem(GameTime gameTime)
        {
            if (EntityManager == null) return;

            int cameraEntityId = EntityManager.CameraEntityId;
            if (cameraEntityId < 0) return;

            Span<TransformComponent> transforms = GetTransformComponentsAsSpan();
            ref readonly TransformComponent camTransform = ref transforms[cameraEntityId];

            float zoom = EntityManager.CameraZoom;
            if (zoom <= 0.001f) zoom = 0.001f;

            float viewportWidth = 1920f;
            float viewportHeight = 1080f;
            Vector2 screenCenter = new Vector2(viewportWidth / 2f, viewportHeight / 2f);
            Vector2 camPos = new Vector2(camTransform.Position.X, camTransform.Position.Y);

            // Simulate view matrix calculation
            ViewMatrix = Matrix.CreateTranslation(-camPos.X, -camPos.Y, 0)
                * Matrix.CreateRotationZ(0f)
                * Matrix.CreateScale(zoom)
                * Matrix.CreateTranslation(screenCenter.X, screenCenter.Y, 0);
        }

        public override void SharedEntityForEachUpdate(ref Yotsuba Entidad, GameTime time) { }
        public override void SharedEntityInitialize(ref Yotsuba Entidad) { }
        public override void Dispose() { }
    }
}

File: Benchmarks_FSharp/Core/Systems/CameraSystem.fs
namespace Benchmarks.Core.Systems

open Benchmarks.Core.Entity
open Benchmarks.Core.Scene
open Benchmarks.Core.Systems.Contract
open Benchmarks.Core.Types

type CameraSystem() =
    inherit ISystem()

    member val ViewMatrix = Matrix.Identity with get, set
    member val ProjectionMatrix = Matrix.Identity with get, set

    override this.InitializeSystem(entities: EntityManager) =
        this.EntityManager <- entities

    override this.UpdateSystem(gameTime: GameTime) =
        if isNull this.EntityManager then ()
        else
        let cameraEntityId = this.EntityManager.CameraEntityId
        if cameraEntityId < 0 then ()
        else
        let transforms = this.GetTransformComponentsAsSpan()
        let camTransform = &transforms.[cameraEntityId]
        let mutable zoom = this.EntityManager.CameraZoom
        if zoom <= 0.001f then zoom <- 0.001f

        let viewportWidth = 1920f
        let viewportHeight = 1080f
        let screenCenter = Vector2(viewportWidth / 2f, viewportHeight / 2f)
        let camPos = Vector2(camTransform.Position.X, camTransform.Position.Y)

        this.ViewMatrix <- Matrix.CreateTranslation(-camPos.X, -camPos.Y, 0f)
                           * Matrix.CreateRotationZ(0f)
                           * Matrix.CreateScale(zoom)
                           * Matrix.CreateTranslation(screenCenter.X, screenCenter.Y, 0f)

    override _.SharedEntityForEachUpdate(_, _) = ()
    override _.SharedEntityInitialize(_) = ()
    override _.Dispose() = ()
